Source/Session.swift

```diff
@@ -469,21 +489,33 @@ open class Session {
     }
 
     func updateStatesForTask(_ task: URLSessionTask, request: Request) {
-        switch (startRequestsImmediately, request.state) {
-        case (true, .initialized):
-            request.resume()
-        case (false, .initialized):
-            // Do nothing.
-            break
-        case (_, .resumed):
-            task.resume()
-            request.didResumeTask(task)
-        case (_, .suspended):
-            task.suspend()
-            request.didSuspendTask(task)
-        case (_, .cancelled):
-            task.cancel()
-            request.didCancelTask(task)
+        requestStateQueue.sync {
+            switch (startRequestsImmediately, request.state) {
```

Comment:
While not super obvious, this is the core issue that is resolved in this PR. Previously, after the `request.state` is read, it could be modified when running the `switch`. This produces all sorts of tricky edge cases that need to be handled. Fundamentally, we need to be able to rely on that state not being swapped out from under us to ensure the operation we're running is still the correct one to run. This ensures that's the case.